pyomo/repn/tests/ampl/test_nlv2.py

```diff
@@ -2308,3 +2361,345 @@ def test_discrete_var_tabulation(self):
+ 2 1	#max name lengths: constraints, variables
+ 0 0 0 1 0	#common exprs: b,c,o,c1,o1
+V1 1 1	#e
+0 1
+n19
+C0	#c1
+o2	#*
+n3
+v1	#e
+x0	#initial guess
+r	#1 ranges (rhs's)
+2 0	#c1
+b	#1 bounds (on variables)
+3	#x
+k0	#intermediate Jacobian column lengths
+J0 1	#c1
+0 1
+""",
+                OUT.getvalue(),
+            )
+        )
+
+        OUT = io.StringIO()
+        nl_writer.NLWriter().write(
+            m,
+            OUT,
+            symbolic_solver_labels=True,
+            linear_presolve=True,
+            export_defined_variables=False,
+        )
+        self.assertEqual(
+            *nl_diff(
+                """g3 1 1 0	# problem unknown
+ 1 1 0 0 0	#vars, constraints, objectives, ranges, eqns
+ 1 0 0 0 0 0	#nonlinear constrs, objs; ccons: lin, nonlin, nd, nzlb
+ 0 0	#network constraints: nonlinear, linear
+ 1 0 0	#nonlinear vars in constraints, objectives, both
+ 0 0 0 1	#linear network variables; functions; arith, flags
+ 0 0 0 0 0	#discrete variables: binary, integer, nonlinear (b,c,o)
+ 1 0	#nonzeros in Jacobian, obj. gradient
+ 2 1	#max name lengths: constraints, variables
+ 0 0 0 0 0	#common exprs: b,c,o,c1,o1
+C0	#c1
+o2	#*
+n3
+o0	#+
+v0	#x
+o54	#sumlist
+3	#(n)
+o2	#*
+n3
+n3
+o5	#^
+n3
+n2
+o3	#/
+n3
+n3
+x0	#initial guess
+r	#1 ranges (rhs's)
+2 0	#c1
+b	#1 bounds (on variables)
+3	#x
+k0	#intermediate Jacobian column lengths
+J0 1	#c1
+0 1
+""",
+                OUT.getvalue(),
+            )
+        )
+
+        OUT = io.StringIO()
+        nl_writer.NLWriter().write(
+            m,
+            OUT,
+            symbolic_solver_labels=True,
+            linear_presolve=False,
+            export_defined_variables=True,
+        )
+        self.assertEqual(
+            *nl_diff(
+                """g3 1 1 0	# problem unknown
+ 3 2 0 0 1	#vars, constraints, objectives, ranges, eqns
+ 1 0 0 0 0 0	#nonlinear constrs, objs; ccons: lin, nonlin, nd, nzlb
+ 0 0	#network constraints: nonlinear, linear
+ 3 0 0	#nonlinear vars in constraints, objectives, both
+ 0 0 0 1	#linear network variables; functions; arith, flags
+ 0 0 0 0 0	#discrete variables: binary, integer, nonlinear (b,c,o)
+ 5 0	#nonzeros in Jacobian, obj. gradient
+ 2 1	#max name lengths: constraints, variables
+ 0 0 0 2 0	#common exprs: b,c,o,c1,o1
+V3 0 1	#nl(e)
+o54	#sumlist
+3	#(n)
+o2	#*
+v0	#y
+v2	#z
+o5	#^
+v0	#y
+n2
+o3	#/
+n3
+v2	#z
+V4 1 1	#e
+1 1
+v3	#nl(e)
+C0	#c1
+o2	#*
+v0	#y
+v4	#e
+C1	#c2
+n0
+x0	#initial guess
+r	#2 ranges (rhs's)
+2 0	#c1
+4 0	#c2
+b	#3 bounds (on variables)
+2 3	#y
+3	#x
+1 3	#z
+k2	#intermediate Jacobian column lengths
+2
+3
+J0 3	#c1
+0 0
+1 1
+2 0
+J1 2	#c2
+0 1
+2 -1
+""",
+                OUT.getvalue(),
+            )
+        )
+
+    def test_presolve_fixes_nl_external_function(self):
+        # This tests a workaround for a bug in the ASL where external
+        # functions with constant argument expressions are not
+        # evaluated correctly.
+        DLL = find_GSL()
```

Comment:
I have a concern about what happens if this doesn't work for some reason, but it's not enough of a concern to stop this PR.